libcaf_io/src/detail/remote_group_module.cpp

```diff
@@ -0,0 +1,153 @@
+  CAF_ASSERT(origin != none);
+  bool lazy_connect = false;
+  auto instance = critical_section([&, this] {
+    if (stopped_) {
+      return group_tunnel_ptr{nullptr};
+    } else {
+      auto& instances = nodes_[origin];
+      if (auto i = instances.find(group_name); i != instances.end()) {
+        return i->second;
+      } else {
+        lazy_connect = true;
+        auto instance = make_counted<detail::group_tunnel>(this, group_name,
+                                                           origin);
+        instances.emplace(group_name, instance);
+        return instance;
+      }
+    }
+  });
+  if (lazy_connect)
+    mm_->resolve_remote_group_intermediary(origin, group_name,
+                                           make_callback(instance));
+  return instance;
+}
+
+group_tunnel_ptr remote_group_module::get_impl(actor intermediary,
+                                               const std::string& group_name) {
+  CAF_ASSERT(intermediary != nullptr);
+  return critical_section([&, this]() {
+    if (stopped_) {
+      return group_tunnel_ptr{nullptr};
+    } else {
+      auto& instances = nodes_[intermediary.node()];
+      if (auto i = instances.find(group_name); i != instances.end()) {
+        auto result = i->second;
+        result->connect(std::move(intermediary));
+        return result;
+      } else {
+        auto result = make_counted<detail::group_tunnel>(
+          this, group_name, std::move(intermediary));
+        instances.emplace(group_name, result);
+        return result;
+      }
+    }
+  });
+}
+
+group_tunnel_ptr remote_group_module::lookup(const node_id& origin,
+                                             const std::string& group_name) {
+  return critical_section([&, this] {
+    if (auto i = nodes_.find(origin); i != nodes_.end())
+      if (auto j = i->second.find(group_name); j != i->second.end())
+        return j->second;
+    return group_tunnel_ptr{nullptr};
+  });
+}
+
+void remote_group_module::drop(const group_tunnel_ptr& instance) {
+  CAF_ASSERT(instance != nullptr);
+  critical_section([&, this] {
+    if (auto i = nodes_.find(instance->origin()); i != nodes_.end()) {
+      if (auto j = i->second.find(instance->identifier());
+          j != i->second.end()) {
+        i->second.erase(j);
+        if (i->second.empty())
+          nodes_.erase(i);
+      }
+    }
+  });
+  instance->stop();
+}
+
+void remote_group_module::connect(const group_tunnel_ptr& instance,
+                                  actor intermediary) {
+  CAF_ASSERT(instance != nullptr);
+  bool stop_instance = critical_section([&, this] {
+    if (auto i = nodes_.find(instance->origin()); i != nodes_.end()) {
+      if (auto j = i->second.find(instance->identifier());
+          j != i->second.end() && j->second == instance) {
+        instance->connect(intermediary);
+        return false;
+      }
+    }
+    return true;
+  });
+  if (stop_instance)
+    instance->stop();
```

Comment:
Why is this not `drop(instance)`? The description of drop reads exactly like this would be the place to call it.

Reply:
No, `drop` removes the instance from the module. Here, `stop_intance` is true if and only if `instance` does not belong to the module. So there's no point in calling `drop` here to remote it from the manager and we instead just make sure to stop the instance before discarding it.